grafana/grafana.py

```diff
@@ -0,0 +1,163 @@
+from redbot.core.utils.chat_formatting import box, humanize_list
+
+
+from .utils import Panel
+
+
+class Grafana(commands.Cog):
+
+    __author__ = "Predä"
+    __version__ = "1.0"
+
+    def __init__(self, bot: Red):
+        self.bot = bot
+        self.session = aiohttp.ClientSession()
+        self.config = Config.get_conf(self, identifier=0xEA016D013C7B488894399820F2BE9874)
+        default_global = {
+            "url": "http://localhost:3000",
+            "dashboard_id": None,
+            "panels": {},
+        }
+        self.config.register_global(**default_global)
+
+    def cog_unload(self):
+        self.bot.loop.create_task(self.session.close())
+
+    def format_help_for_context(self, ctx: commands.Context) -> str:
+        """Thanks Sinbad!"""
+        pre_processed = super().format_help_for_context(ctx)
+        return f"{pre_processed}\n\nAuthor: {self.__author__}\nCog Version: {self.__version__}"
+
+    async def fetch_grafana(self, **kwargs):
+        from_time = kwargs.get("timedelta")
+        params = {
+            "orgId": 1,
+            "from": int((datetime.now() - from_time).timestamp()) * 1000,
+            "to": int(time.time()) * 1000,
+            "panelId": kwargs.get("panelid"),
+            "width": 1000,
+            "height": 500,
+        }
+        async with self.bot.session.get(
+            f"{await self.config.url()}/render/d-solo/" f"{await self.config.dashboard_id()}",
+            params=params,
+        ) as resp:
+            if resp.status != 200:
+                return None, {}
+            return BytesIO(await resp.read()), params
+
+    @commands.group(invoke_without_command=True)
+    @commands.cooldown(1, 2, commands.BucketType.user)
+    @commands.max_concurrency(1, commands.BucketType.guild)
+    async def graph(
+        self,
+        ctx: commands.Context,
+        panel: Panel,
+        from_time: TimedeltaConverter = timedelta(days=1),
+    ):
+        """Render an image of a selected panel of [botname] metrics."""
+        async with ctx.typing():
+            file, params = await self.fetch_grafana(panelid=panel.id, timedelta=from_time)
+        if not file:
+            return await ctx.send("Failed to fetch Grafana.")
+        msg = (
+            f"{await self.config.url()}/d/{await self.config.dashboard_id()}?"
+            f"panelId={params['panelId']}&fullscreen&orgId={params['orgId']}"
+            f"&from={params['from']}&to={params['to']}&refresh=15m"
+        )
+        filename = "&".join([f"{k}={v}" for k, v in params.items()])
+        return await ctx.send(msg, file=discord.File(file, filename=f"graph-{filename}.png"))
+
+    @graph.command(name="list")
+    async def list_graphs(self, ctx: commands.Context):
+        """List all panels that can be used with `[p]graph` command."""
+        await ctx.send(box(humanize_list(list((await self.config.panels()).keys()))))
+
+    @graph.group(name="set")
+    @commands.is_owner()
+    async def set_graphs(self, ctx):
```

Comment:
It's missing `commands.Context` type hint for this command group and its subcommands.